### ISSUE TITLE
Key error when serializing

### ISSUE THREAD

**Issue by bradhackinen** [spacy-io/spaCy#350]
I received the following error when trying to serialize a document:

> Traceback (most recent call last):
>   File "generateSpaCyDocs.py", line 41, in <module>
>     f.write(doc.to_bytes())
>   File "spacy/tokens/doc.pyx", line 425, in spacy.tokens.doc.Doc.to_bytes (spacy/tokens/doc.cpp:1056
> 6)
>     byte_string = self.vocab.serializer.pack(self)
>   File "spacy/serialize/packer.pyx", line 110, in spacy.serialize.packer.Packer.pack (spacy/serializ
> e/packer.cpp:5683)
>     codec.encode(array[:, i], bits)
>   File "spacy/serialize/huffman.pyx", line 61, in spacy.serialize.huffman.HuffmanCodec.encode (spacy
> /serialize/huffman.cpp:2535)
>     i = self._map[word]
> KeyError: 11222

The raw text might have some pretty weird characters and tokens in it. Do I need to drop all out-of-vocabulary tokens from the doc before serializing? I'm a little fuzzy on how these are handled.

**Comment by honnibal** [comment]
You shouldn't have to drop anything from the text before serialising --- it should just work.

Can you post the text that caused this?

**Comment by bradhackinen** [comment]
[com000031.PDF.txt](https://github.com/spacy-io/spaCy/files/239555/com000031.PDF.txt)

This file will generate the error with the following code:
```
from spacy import English
from HTMLParser import HTMLParser

nlp = English(tagger=False)
parser = HTMLParser()

testFile = 'com000031.PDF.txt'

with open(testFile,'rb') as f:
    doc = nlp(parser.unescape(f.read().decode('utf8')))

doc.to_bytes()
```
It looks like the problem is caused by trying to serialize a doc that is not parsed with the tagger. When I load English as `nlp = English()`, the error disappears.

For other documents, I can serialize them without the tagger or parser, but then they fail to load properly.

Maybe this is as designed (I can see how it might be necessary), but it would be nice to be able to skip the parsing or tagging and still have the serializer work.

SpaCy is amazing overall though. Thanks for sharing it.

**Comment by wbwseeker** [comment]
Hey,

@honnibal, I found this too last week. Back then, I wrote some tests to test this, I've uploaded it to the repository.
run `python -m pytest spacy --models -k test_serialization` to see the output

It is as @bradhackinen says, if the pos tags aren't set, the deserialization fails.
The reason seems to be that during deserialization, this line (https://github.com/spacy-io/spaCy/blob/master/spacy/tokens/doc.pyx#L394) fails because morphology.reverse_index is empty.
The tag values are all set to 0 as they should but assign_tag doesn't get a proper value from reverse_index for it.
